mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -745,63 +746,61 @@ struct AttentionRewritePattern : public OpRewritePattern<tosa::MatMulOp> {
   }
 
   FailureOr<Value> maybeSoftmaxDenominator(Value val) const {
-    tosa::ReduceSumOp rsum = getDefiningNonReshapeOp<tosa::ReduceSumOp>(val);
+    auto rsum = getDefiningNonReshapeOp<tosa::ReduceSumOp>(val);
     if (!rsum) {
       return failure();
     }
     return maybeSoftmaxNumerator(rsum.getInput());
   }
 
   FailureOr<Value> maybeSoftmax(Value val) const {
-    tosa::MulOp mul = getDefiningNonReshapeOp<tosa::MulOp>(val);
+    auto mul = getDefiningNonReshapeOp<tosa::MulOp>(val);
     if (!mul) {
       return failure();
     }
-    if (tosa::ReciprocalOp rec =
+    if (auto rec =
             getDefiningNonReshapeOp<tosa::ReciprocalOp>(mul.getInput1())) {
       return maybeSoftmaxDenominator(rec.getInput1());
-    } else if (tosa::ReciprocalOp rec =
-                   getDefiningNonReshapeOp<tosa::ReciprocalOp>(
-                       mul.getInput2())) {
+    } else if (auto rec = getDefiningNonReshapeOp<tosa::ReciprocalOp>(
+                   mul.getInput2())) {
       return maybeSoftmaxDenominator(rec.getInput1());
     } else {
       return failure();
     }
   }
 
-  FailureOr<std::tuple<tosa::MatMulOp, TypedValue<TensorType>>>
-  getMatMulAndScaleInputs(tosa::MulOp scale) const {
-    if (tosa::MatMulOp mm =
-            getDefiningNonReshapeOp<tosa::MatMulOp>(scale.getInput1())) {
-      return std::make_tuple(mm, scale.getInput2());
+  template <typename TargetType, typename RootType>
+  FailureOr<std::tuple<TargetType, TypedValue<TensorType>>>
+  decomposeOp(RootType root) const {
```

Comment:
Naively, rootType feels like the root you find through traversals up via getDefiningNonReshapeOp.

So before, this function's purpose was self explanatory : that is get the matmul and the scale input.
Now, decomposeOp with TargetType and RootType is not self explanatory  to me.

Can you clarify your thinking behind the naming ?

Reply:
Hi @manupak,

I wanted to re-use your code and I had to come up to some generic name.

Let's suppose we have a sub-expression inside the attention as follows:

```
%1 = tosa.matmul (%arg0, %arg1)
%2 = tosa.mul(%1, %arg3)
```

```
              MulOp (Root)
             /         \
            /           \
        MatMulOp        %arg3
        (Target)
           /   \
      %arg0    %arg1
```

So, the `decomposeOp` takes an Op as the root (MulOp), checks whether it was composed with an op of `TargetType` and `TensorType`. If yes, it will returns the operands of the root Op.

Probably, the name is the issue. I would appreciate if you can suggest a better name.

Here is another example where this method is useful. Let's conside

```
%1 = tosa.matmul (%arg0, %arg1)
%2 = tosa.mul(%1, %arg3)
%3 = tosa.add(%2, %arg4)
```

```
                   AddOp
                   (Root)
                  /      \
                 /         \           
              MulOp       %arg4
             (Target)
             /         \
            /           \
        MatMulOp        %arg3
           /   \
      %arg0    %arg1
```

Reply:
How about getOperandsOrdered(...) ?
with RootType -> currOpType
with TargetType -> FirstOperandType

Reply:
Makes sense